Uniffi non-flat enums are not forward-declared

Uniffi non-flat enums are not forward-declared in the generated C++ header, which can cause compilation errors.

```rust
use std::sync::Arc;

uniffi::setup_scaffolding!();

#[derive(Clone, Debug, uniffi::Object)]
pub struct Foo {
    x: i32,
}

#[uniffi::export]
impl Foo {
    #[uniffi::constructor]
    pub fn new(_b: Bar) -> Arc<Self> {
        Arc::new(Foo { x: 33 })
    }
}

#[derive(Clone, Debug, uniffi::Enum)]
pub enum Bar {
    A { x: i32 },
    B { x: i32 },
}
```

The generated header contains these forward declarations and definitions in this order:
```cpp
struct Foo;

namespace uniffi {
    struct FfiConverterFoo;
} // namespace uniffi

struct Foo {
    friend uniffi::FfiConverterFoo;

    Foo() = delete;

    Foo(const Foo &) = delete;
    Foo(Foo &&) = delete;

    Foo &operator=(const Foo &) = delete;
    Foo &operator=(Foo &&) = delete;

    ~Foo();
    static std::shared_ptr<Foo> init(Bar _b);

private:
    Foo(void *);

    void *instance;
};

namespace uniffi {
struct FfiConverterTypeBar;
} // namespace uniffi

struct Bar {
    friend uniffi::FfiConverterTypeBar;
    struct kA {
        int32_t x;
    };
    struct kB {
        int32_t x;
    };
    Bar(kA variant): variant(variant) {}
    Bar(kB variant): variant(variant) {}

    Bar(const Bar &other): variant(other.variant) {}
    Bar(Bar &&other): variant(std::move(other.variant)) {}

    Bar &operator=(const Bar &other) {
        variant = other.variant;
        return *this;
    }

    Bar &operator=(Bar &&other) {
        variant = std::move(other.variant);
        return *this;
    }

    /**
     * Returns the variant of this enum
     */
    const std::variant<kA, kB> &get_variant() const {
        return variant;
    }

private:
    std::variant<kA, kB> variant;

    Bar();
};
```

Attempting to compile a C++ file including this header triggers the following error with clang++:
```
In file included from foobar.cpp:1:
./unifficppbug.hpp:38:38: error: unknown type name 'Bar'
   38 |     static std::shared_ptr<Foo> init(Bar _b);
      |                                      ^
1 error generated.
```

--- a/bindgen/src/bindings/cpp/templates/wrapper.hpp
+++ b/bindgen/src/bindings/cpp/templates/wrapper.hpp
@@ -30,14 +30,12 @@ namespace {{ namespace }} {
 {%- match typ %}
 {%- when Type::Enum { module_path, name } %}
 {%- let e = ci|get_enum_definition(name) %}
-{%- if ci.is_name_used_as_error(name) %}
+{%- if ci.is_name_used_as_error(name) || !e.is_flat() %}
 struct {{ name }};
 {%- else %}
-{%- if e.is_flat() %}
 enum class {{ name }};
 {%- endif %}
-{%- endif %}
-{%- when Type::Record { module_path, name } %} 
+{%- when Type::Record { module_path, name } %}
 struct {{ name }};
 {%- when Type::Object { module_path, name, imp } %}
 struct {{ name }};

